src/test/java/com/palantir/gradle/versions/CheckOverbroadConstraintsTest.java

```diff
@@ -0,0 +1,322 @@
+                .orElseThrow(() -> new RuntimeException(
+                        String.format("%s found in file names but could not retrieve the file.", testCase.fileName)));
+
+        String actualContent =
+                Files.readString(resourceFile.toPath(), StandardCharsets.UTF_8).trim();
+
+        assertThat(actualContent)
+                .as(
+                        "Contents of file '%s' do not match the expected test case. Run tests locally to fix.",
+                        testCase.fileName)
+                .isEqualTo(testCase.fileContents().trim());
+    }
+
+    private static void deleteOldTests(Path path) throws IOException {
+        if (!Files.exists(path)) {
+            return;
+        }
+        try (DirectoryStream<Path> stream = Files.newDirectoryStream(path)) {
+            for (Path entry : stream) {
+                if (Files.isDirectory(entry)) {
+                    deleteOldTests(entry);
+                    Files.delete(entry);
+                } else {
+                    Files.delete(entry);
+                }
+            }
+        }
+    }
+
+    public static final class TestCase {
+        private final String fileName;
+        private final VersionsProps versionsProps;
+        private final LockState lockState;
+        private final List<String> propsLines;
+        private final String lockLines;
+
+        private TestCase(Builder builder) {
+            this.fileName = builder.fileName;
+            this.versionsProps = builder.versionsProps;
+            this.lockState = builder.lockState;
+            this.propsLines = builder.propsLines;
+            this.lockLines = builder.lockLines;
+        }
+
+        public static Builder builder(String testName) {
+            return new Builder(testName);
+        }
+
+        public List<String> newPropsLines() {
+            Map<String, List<String>> oldToNewLines =
+                    CheckOverbroadConstraints.determineNewLines(versionsProps, lockState);
+            return CheckOverbroadConstraints.generateUpdatedPropsLines(oldToNewLines, propsLines);
+        }
+
+        public String fileContents() {
+            Patch<String> patch = DiffUtils.diff(propsLines, newPropsLines());
+
+            String diff =
+                    DiffUtils.generateUnifiedDiff(fileName, fileName, propsLines, patch, Integer.MAX_VALUE).stream()
+                            // Skip the unified diff header
+                            .skip(3)
+                            .collect(Collectors.joining("\n"));
+
+            if (diff.isEmpty()) {
+                diff = "<unchanged>";
+            }
+
+            return String.join(
+                    "\n",
+                    "versions.props",
+                    String.join("\n", propsLines),
+                    "",
+                    "versions.lock:",
+                    lockLines,
+                    "",
+                    "versions.props diff:",
+                    diff,
+                    "");
```

Comment:
I added this to make the diffs nicer. But the current code is always stripping the newline at the end of the versions.props if there is one. We should make sure to preserve the newline if this is there or not. Might be worth writing this as a check that runs on every single test rather than depending on PR review of the examples.

Reply:
So I think what is happening here is that diff utils is removing the new line as it is shared between the old and new versions props - what I have done is add some specific test case that only check generateUpdatedPropsLines they should cover checking new lines and insertion order

Reply:
It's removing the newline unnecessarily on internal projects too. Trust me, angery people will appear in our support with we excavate out newline changes... (I have done this many times).

Reply:
ah! The culprit was the read from file function not adding the new line to the list - so the test I added wouldn't catch it - Ive modified the groovy tests to include checking for new line